### DISCUSSION TITLE
Let material properties of non-mentioned blocks equal to 0

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#19470] [Q&A General] [answered]
Hi guys,

is there a cleaner way to let material properties of non-mentioned blocks equal to 0, I know the messy way (shown below)

```
  [./excav_perm1]
    type = GenericConstantMaterial
    prop_names  = excav_perm
    prop_values = 1
    block = 1
  [../]

  [./unmined_perm1]
    type = GenericConstantMaterial
    prop_names  = excav_perm
    prop_values = 0
    block = '2 3 4 5 6'
  [../]
```

but is it possible to just define the property on a particular block(s), OR define 2 values in a single object for example
 ```
 [./excav_perm1]
    type = GenericConstantMaterial
    prop_names  = excav_perm
    prop_values_1 = 1
    blocks_prop_1 = 1

    prop_values_2 = 0
    blocks_prop_2 = '2 3 4 5 6'
  [../]
```

Not sure where to start, should it be somewhere in the .C file?

Kind regards,
Traiwit

**Comment by GiudGiud** [comment, marked as answer]
Hi Traiwit

Is this what you're looking for
https://mooseframework.inl.gov/source/materials/PiecewiseConstantByBlockMaterial.html

Guillaume

**Reply by GiudGiud** [reply]
Is the material used in all blocks btw? When not used you can just omit it.

**Reply by aprilnovak** [reply]
You might need to set `material_coverage_check = false` in the `[Problem]` block if by omitting your `[unmined_perm1]` block you wouldn't otherwise have any materials at all on blocks `2 3 4 5 6`.

**Reply by Traiwit** [reply]
Hi @GiudGiud, sorry for the late reply, I was busy with other stuff early today.

Is this a new material object? I couldn't find it in my MOOSE dir, I try to look for an example now sure how to input `subdomain_to_prop_value`

![image](https://user-images.githubusercontent.com/79562439/143382791-eecff7fa-f8ce-45eb-abd3-8a7b71715c99.png)

so in this case, if I have 100+ blocks, I have to list all of them here, right?

@aprilnovak noted, i learned that when I first try to use [Controls] with [Materials], it ignores the duplicated materials, but it doesn't ignore the unassigned materials, unfortunately.

> Is the material used in all blocks btw? When not used you can just omit it.

no, I actually want to only assign the properties to the mined blocks, and the unmined blocks the properties would be 0

Kind regards,
traiwit

**Reply by GiudGiud** [reply]
It s not new but I renamed it recently. In your version MOOSE it s still call FVPropValPerSubdomain or something like that

You have to list all the ones you want to define a value for yes

**Reply by Traiwit** [reply]
Hi @GiudGiud 

it doesnt seem to work from my end, not sure if I got the right syntax?

 ```
 [./excav_poro1]
    type = FVPropValPerSubdomainMaterial
    prop_name  = 'excav_poro'
    subdomain_to_prop_value =  'ALL_TET10 1
                                DRIVE1_TET10 1
                                DRIVE2_TET10 1
                                DRIVE3_TET10 1
                                DRIVE4_TET10 1
                                FLOOR_TET10 1
                                STOPE2_TET10 1
                                STOPE3_TET10 1
                                STOPE4_TET10 1
                                WALLS_TET10 1'
  [../]
```

AuxVar/AuxKernel couldn't pick it up

```
[./excav_poro]
order = CONSTANT
family = MONOMIAL
[../]
[./excav_poro]
    type = MaterialRealAux
    property = excav_poro
    variable = excav_poro
  [../]
```

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
I think FVPropValPerSubdomainMaterial only works with FV kernels

just wondering if PiecewiseConstantByBlockMaterial has been merged yet?

**Reply by GiudGiud** [reply]
There s no distinction between FV and FE kernels for auxkernels, because there is no (face) flux kernel for the auxiliary system.

PiecewiseConstantByBlockMaterial is indeed merged but it s just a rename so that wont help you. 

What's the error message you get when you say it cant pick it up

**Reply by Traiwit** [reply]
@GiudGiud, my bad I thought I pasted it earlier

```
Material property 'excav_poro', requested by 'excav_poro' is not defined on block ALL_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block DRIVE1_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block DRIVE2_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block DRIVE3_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block DRIVE4_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block FLOOR_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block STOPE2_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block STOPE3_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block STOPE4_TET10
Material property 'excav_poro', requested by 'excav_poro' is not defined on block WALLS_TET10
```

**Reply by GiudGiud** [reply]
Could be an AD thing, like FVPropVal may be defining AD properties
I ll check later.

**Reply by Traiwit** [reply]
Noted, Thanks @GiudGiud

**Reply by GiudGiud** [reply]
ok it s not an AD thing. We also moved this material to be a functor material property instead of a material and did not (re)rename it.

If you want to use it you ll have to use its old definition, as a material property. Here are the files:
https://github.com/GiudGiud/moose/blob/PR_doco8/framework/src/materials/FVPropValPerSubdomainMaterial.C
https://github.com/GiudGiud/moose/blob/PR_doco8/framework/include/materials/FVPropValPerSubdomainMaterial.h

**Reply by Traiwit** [reply]
Hi @GiudGiud I'm a bit confused, so what would be the use case for `PiecewiseConstantByBlockMaterial` then?

**Reply by GiudGiud** [reply]
The current code is made for Navier Stokes finite volume, and more generally anyone who would want to use functor material properties. The functor system allows us to seamlessly swap between functions, material properties and variables, with unified parameters. It was mostly added to do local on-the-fly evaluations (eg face evaluations)

Old material properties, the files i linked, work a little differently, they get computed before the kernels are run on an element, when the element is `reinit`-ialized.

**Reply by GiudGiud** [reply]
i ll will be re-adding the PiecewiseConstantByBlockMaterial for regular material properties soon and renaming the current PiecewiseConstantByBlockMaterial to PiecewiseByBlockFunctorMaterial.

In the mean time feel free to use those two files I linked to

**Reply by Traiwit** [reply]
@GiudGiud awesome, thank you! please keep me updated

**Reply by Traiwit** [reply]
I tried `FVPropValPerSubdomainMaterial.C` it still wouldn't work

```
[./permeability_1]
    type = PorousFlowPermeabilityConstCSV
    read_prop_user_object = perm_read
    excav_perm = excav_perm
  [../]

  [./excav_poro1]
    type = FVPropValPerSubdomainMaterial
    prop_name  = excav_perm
    subdomain_to_prop_value =  'ALL_TET10 0
                                DRIVE1_TET10 0
                                DRIVE2_TET10 0
                                DRIVE3_TET10 0
                                DRIVE4_TET10 0
                                FLOOR_TET10 0
                                STOPE2_TET10 1e-2
                                STOPE3_TET10 1e-2
                                STOPE4_TET10 1e-2
                                WALLS_TET10 0'
  [../]
```

```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'excav_perm', requested by 'permeability_1' is not defined on block ALL_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block ALL_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block ALL_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block DRIVE1_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block DRIVE1_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block DRIVE1_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block DRIVE2_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block DRIVE2_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block DRIVE2_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block DRIVE3_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block DRIVE3_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block DRIVE3_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block DRIVE4_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block DRIVE4_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block DRIVE4_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block FLOOR_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block FLOOR_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block FLOOR_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block STOPE2_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block STOPE2_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block STOPE2_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block STOPE3_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block STOPE3_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block STOPE3_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block STOPE4_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block STOPE4_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block STOPE4_TET10
Material property 'excav_perm', requested by 'permeability_1' is not defined on block WALLS_TET10
Material property 'excav_perm', requested by 'permeability_1_face' is not defined on block WALLS_TET10
Material property 'excav_perm', requested by 'permeability_1_neighbor' is not defined on block WALLS_TET10
```
BUT when I do it the messy way (below), it works fine

```
  [./excav_perm3]
    type = GenericConstantMaterial
    prop_names  = excav_perm
    prop_values = 10
    block =  'STOPE2_TET10 STOPE3_TET10 STOPE4_TET10'
  [../]
  
  [./unmined_perm3]
    type = GenericConstantMaterial
    prop_names  = excav_perm
    prop_values = 0
    block =                       'ALL_TET10
                                  DRIVE1_TET10
                                  DRIVE2_TET10
                                  DRIVE3_TET10
                                  DRIVE4_TET10
                                  FLOOR_TET10
                                  WALLS_TET10'
  [../]
```
but that's okay though, I just want to avoid having too many objects in the input file, for some reason, it slows down the problem-setting stage of the simulation

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
weird. 
did you use the file I linked to?
~~Are you using AD in your simulations?~~ nvmd about that one, you are not since GenericConstantMaterial worked

I have a PR up that will replace FVPropValPerSubdomain with PiecewiseByBlockMaterial, it s not merged yet though

**Reply by Traiwit** [reply]
hi @GiudGiud, yes i use those files you provided

**Reply by GiudGiud** [reply]
@Traiwit my PR got merged can you please try PiecewiseByBlockConstantMaterial ? (it s in moose/next, not in master yet, maybe tomorrow)

**Reply by Traiwit** [reply]
@GiudGiud it works fine now, awesome, thank you!
